### PR TITLE
Refactor copybooks processing

### PR DESCRIPTION
**Pull request by temanbrcom**

### CODE REVIEW

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+      ExtendedDocument copybookDocument =
+          buildExtendedDocumentForCopybook(metaData).apply(hierarchy).unwrap(errors::addAll);
+      return stack -> {
+        writeText(metaData, copybookDocument).accept(stack);
+        return subContext -> {
+          storeCopyStatementSemantics(metaData, copybookDocument).accept(subContext);
+          return nestedMappings -> {
+            collectNestedSemanticData(metaData, copybookDocument).accept(nestedMappings);
+            return allErrors -> allErrors.addAll(errors);
```
Comment (grianbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java:113]: Oh my... Is it legal?

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+          };
+        };
+      };
+    };
+  }
+
+  private Function<CopybookMetaData, ResultWithErrors<CopybookMetaData>> validateMetaData(
+      int maxCopybookNameLength) {
```
Comment (grianbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java:121]: No way, man. You got `maxCopybookNameLength` as a private final field. There is no need to construct a new function.

**Review by grianbrcom — COMMENTED** [review]
17 from 36 passed, I can't stand more for today.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2021 Broadcom.
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java:2]: Oops, seems like it is automatic. I will fix it.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  /**
+   * Replace the text using the internal state of the hierarchy
+   *
+   * @param text text to replace
+   * @param accumulator a function for applying the replacing
+   * @return replaced text
+   */
+  public String replaceText(
+      String text, BiFunction<String, List<Pair<String, String>>, String> accumulator) {
+    return recursiveReplaceStmtStack.stream().reduce(text, accumulator, (raw, res) -> res);
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java:142]: Good catch. It is just the inheritance from the previous implementation. I will fix it.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/TextPreprocessor.java [hunk 5]
```diff
@@ -37,28 +33,21 @@
...
   /**
-   * Checks and clean of the code as per cobol program structure.
-   * Like
-   * - checking code length doesn't exceeds 80 char
-   * - removing line number and sequence number from source.
-   * - concatenating continued lines
-   * - Normalizing files in case of any compiler directives
+   * Checks and clean of the code as per cobol program structure. Like - checking code length
+   * doesn't exceeds 80 char - removing line number and sequence number from source. - concatenating
+   * continued lines - Normalizing files in case of any compiler directives
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/TextPreprocessor.java]: Done

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java (hunk 2 above)
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java:121]: Okay...

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java (hunk 1 above)
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/analysis/AbstractCopybookAnalysis.java:113]: Why not? Seems pretty nice.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/GrammarPreprocessorListenerImpl.java [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+  @Override
+  public void exitControlDirective(ControlDirectiveContext ctx) {
+    pop();
+    if (ctx.controlOptions().isEmpty()) reportInvalidArgument(ctx.controlCbl());
+    accumulateTokenShift(ctx);
+  }
+
+  @Override
+  public void exitLinkageSection(LinkageSectionContext ctx) {
+    analyzeCopybook(PREDEFINED, ctx, ctx);
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/GrammarPreprocessorListenerImpl.java:168]: Yes

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/PreprocessorStack.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Broadcom.
+ * Copyright (c) 2021 Broadcom.
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/copybooks/PreprocessorStack.java]: Seems like it is automatic, I will fix it.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java:142]: Well, not exactly. This is actually used to allow the accumulator to apply to the target of a different type, String in this case. There is no such overload for the `reduce` without the combiner.

**Review by grianbrcom — APPROVED** [review]
Good, but it will be better to fix comments.
And I still have no idea why do you have `recursiveReplaceStmtStack`.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/CopybookHierarchy.java]: Of course

File: server/src/main/java/org/eclipse/lsp/cobol/core/semantics/NamedSubContext.java [hunk 8]
```diff
@@ -18,43 +18,70 @@
...
  * This class represents a semantic sub-context for specific types of elements that have names and
  * stores their definitions and variables.
  */
 @Value
-public class NamedSubContext implements SubContext<String> {
-
+public class NamedSubContext {
   Multimap<String, Location> definitions = HashMultimap.create();
   Multimap<String, Location> usages = HashMultimap.create();
+  Map<String, Locality> definitionStatements = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/semantics/NamedSubContext.java:35]: This one is for statements, not for names. I use to escalate the syntax errors through the copybook hierarchy.

**Pull request merged**